docs/pages/guides/config-plugins.md

```diff
@@ -0,0 +1,614 @@
+
+- `modResults`: The object to modify and return. The type depends on the mod that's being used.
+- `modRequest`: Additional properties supplied by the mod compiler.
+  - `projectRoot: string`: Project root directory for the universal app.
+  - `platformProjectRoot: string`: Project root for the specific platform.
+  - `modName: string`: Name of the mod.
+  - `platform: ModPlatform`: Name of the platform used in the mods config.
+  - `projectName?: string`: iOS only: The path component used for querying project files. ex. `projectRoot/ios/[projectName]/`
+
+## Creating a mod
+
+Say you wanted to write a mod to update the Xcode Project's "product name":
+
+```ts
+import { ConfigPlugin, withXcodeProject } from '@expo/config-plugins';
+
+const withCustomProductName: ConfigPlugin = (config, customName) => {
+  return withXcodeProject(config, async config => {
+    // config = { modResults, modRequest, ...expoConfig }
+
+    const xcodeProject = config.modResults;
+    xcodeProject.productName = customName;
+
+    return config;
+  });
+};
+
+// 💡 Usage:
+
+/// Create a config
+const config = {
+  name: 'my app',
+};
+
+/// Use the plugin
+export default withCustomProductName(config, 'new_name');
+```
+
+### Experimental functionality
+
+Some parts of the mod system aren't fully flushed out, these parts use `withDangerousModifier` to read/write data without a base mod. These methods essentially act as their own base mod and cannot be extended. Icons for example, currently use the dangerous mod to perform a single generation step with no ability to customize the results.
+
+```ts
+export const withIcons: ConfigPlugin = config => {
+  return withDangerousModifier(config, async config => {
+    // No modifications are made to the config
+    await setIconsAsync(config, config.modRequest.projectRoot);
+    return config;
+  });
+};
+```
+
+Be careful using `withDangerousModifier` as it is subject to change in the future.
+The order with which it gets executed is not reliable either.
+Currently dangerous mods run first before all other modifiers, this is because we use dangerous mods internally for large file system refactoring like when the package name changes.
+
+## Plugin module resolution
+
+The strings passed to the `plugins` array can be resolved in a few different ways.
+
+> Any resolution pattern that isn't specified below is unexpected behavior, and subject to breaking changes.
+
+### Project file
+
+You can quickly create a plugin in your project and use it in your config.
+
+- ✅ `'./my-config-plugin'`
+- ❌ `'./my-config-plugin.js'`
+
+```
+╭── app.config.js ➡️ Expo Config
+╰── my-config-plugin.js ➡️ ✅ `module.exports = (config) => config`
+```
+
+### app.plugin.js
+
+Sometimes you want your package to export React components and also support a plugin, to do this, multiple entry points need to be used (because the transpilation (Babel preset) may be different).
```

Comment:
```suggestion
Sometimes you want your package to export React components and also support a plugin. To do this, multiple entry points need to be used because the transpilation (Babel preset) may be different.
```